### ISSUE TITLE
How to differentiate networked uploads from the new SG software from manual uploads by a user?

### ISSUE THREAD

**Issue by j-sayers** [MotusDev/Motus-TO-DO#1267]
The new SG software uploads data differently than the old one, essentially like a manual upload instead of as a syncReceiver, and authenticated using valid Motus credentials that are entered during SG setup.

There's nothing particularly wrong with this, except that in the sgdata jobs these appear like any other manual upload and not as a syncReceiver. So they can't be sorted or filtered by upload type, or separated from any other upload that the user's creds are associated with. Which makes a lot of troubleshooting etc very difficult.

@tve said he can easily add something to the HTTP request that identifies these are networked uploads (eg syncReceivers) but I'm not sure if we have anything in place yet to handle that. @denislepage or @Hussell do you have any thoughts about this?

**Comment by tve** [comment]
I'm wondering whether it would be best to funnel uploads through the sensorgnome server where all the monitoring data already goes. That would make it easier to detect when users change sd-cards, move sd-cards from one SG to another, and other stuff like that. In that case, it would be best if there was one set of creds owned by the sensorgnome site that could upload for any project. That would be more similar to the current set-up (although that's a pull model). It would eliminate the need for users to put creds on the SGs, which would be a good thing.

**Comment by denislepage** [comment]
It might be useful to chat on the architecture set up. Setting up sensorgnome as an intermediary in some ways creates one more opportunity for breaking the flow. There is merit to that approach though, at least when it comes to authentication. With CTT, we are using a shared AWS account to exchange files back and forth, which is another possible way to handle file exchanges between sensorgnome and Motus/sgdata, but this adds yet another layer.

In any solution, the authentication problem is the most challenging. We want to make it easy/convenient, but not also prevent misuse (intentional or not).

TBH, I’m still on the fence regarding the utility for sensor gnome to be totally separated as it is now. In many ways, I’d almost prefer sensorgnome.org to be part of our existing network, or even be fully integrated into our platform, so data files and monitoring data would all only have 1 place to go (as well as users). The potential for sensorgnome to be used to sensors other than Motus remains hypothetical and may never be realized. As a minimum, we’d likely want to keep the door open for someone else to use the bones of the software, but they’d likely need to set up their own copy on their server anyway, customized for their needs.

**Comment by tve** [comment]
The reason the sensorgnome server is not currently an intermediary in the upload is that I didn't want to introduce another hop to manage. However, I'm realizing that having user creds out in the field on SGs isn't a good long term solution and there has been no movement to support anything else on the back-end.

I believe the most user-friendly option is to be able to just set-up the SG and then worry about registering it later. The less one has to do over a phone connected to a hot-spot wifi the better. That also makes it easier for those that install many SGs. This approach would imply that the SG connects to the sensorgnome server, provides some secret info and then the server can let users "claim" new or altered sensorgnomes. (This is actually somewhat similar to the current sensorgnome server, except that the latter just accepts any new SG and doesn't have any protection, such as users logging in and explicitly claiming a SG or changes to a SG.)

The reason for a separate server at this point is that it allows for rapid development and experimentation with different solutions.

**Comment by tve** [comment]
I've been poking around a bit more... Some things I observed:
- when one logs into motus.org one gets a cookie with a session token which appears to be valid for a bit over a month
- to upload, a SG really just needs this session token, not a username/password, that makes it a decent time-limited authorization token
- I created a "sg-uploader" user, added him to my project, and gave just data-update permissions:
![image](https://user-images.githubusercontent.com/39480/195761147-e9c08770-e95d-449b-9810-4c13cee4aa43.png)
These are the minimal permissions that allow Mr. sg-uploader to perform uploads

So all this opens up some possibilities. Basically, when a new SG comes up, the sensorgnome server can perform some checks, do a login to motus.org using sg-uploader, and then hand the SG the resulting session token. This allows the SG to upload stuff directly to motus.org for about a month. Before the expiration, the SG has to come back and request a fresh token, which repeats the process. At a high level, that's basically what we're looking for. Now as-is this has some issues:

1. Either project PIs have to add the sg-uploader user to their project or they have to create their own similar user and then enter the creds into the sensorgnome server. Neither option is great from a user point of view.
2. The token handed to the SG allows uploading into any project to which Mr sg-uploader has been added. The token should really only allow uploading to the specific project to which the SG belongs.
3. The "data update" permission is too broad, it allows more stuff than just uploading sensorgnome data. The token should only allow the `/data/project/sgJobs` requests needed to upload data.

These are clear issues, however, the scheme can work today with no changes and it's basically what we're looking for IMHO. The issues can be fixed and it should be able to happen without even changing the SG software as long as all that's needed is a session token in the cookie.

I don't know anything about the auth back-end so I can't really make a lot of suggestions for how to address the issues. One option is that the sg-uploader could be added to all projects in the back-end and perhaps have no permissions at all and the server handler for `/data/project/sgJobs` could special-case the sg-uploader user. This doesn't address issue #2, however, for that the session info referenced by the token would need to encode a project and I don't know whether that's currently possible.

**Comment by tve** [comment]
Quick update: I did implement a scheme where the SG obtains a session_token from the sensorgnome server and then uses that upload. Works like a charm, but I have to test it over the course of a couple of days to make sure renewal and such works properly.
Also, I realized that one can upload data for any SG... So currently I'm using user sg-uploader who only has access to acct 460. If I upload data for a SG in another project (I've abused SG-68B4RPI33F06 in project 1 a bit, Josh uses it for testing) then the data is processed but marked as "there is no deployment in this project for this receiver". (Which is good!)
This opens up another short-term solution for all this, which is to hard-code in sgdata to allow Mr sg-uploader to upload for any project. Not pretty but might be a good stop-gap if a real solution is going to take a long time to devise and implement.

**Comment by j-sayers** [comment]
Good questions. Uploading data via sg.org as currently happens is one solution to the credentials issue (and identifying networked uploads as such). But both @tve and @denislepage have pointed out different drawbacks to that.

The short term solution of hard-coding Mr SG-Uploader would definitely work. Not necessarily pretty but functional!

Aside from Denis, maybe @Hussell and @cjardine-bsc have some ideas of what other options might be possible on the back end, or whether a hard-coded limited-privilege user is the best option for the time being

**Comment by tve** [comment]
BTW, SQLite has glob()... and X GLOB Y...

**Comment by Hussell** [comment]
> I'm trying to understand what this means... Does it mean that now there is absolutely no place to check what happened to a upload job? So troubleshooting impossible or checking whether data has been processed impossible?

No, filtering out automated uploads is optional. It only means that this particular new type of automated upload is now *also* being filtered out when that option is requested.

**Comment by Hussell** [comment]
> BTW, SQLite has glob()... and X GLOB Y...

Thanks, that's very helpful. Though the obvious question is, if they support that, why don't they support the standard SQL syntax for it? A quick search didn't reveal any other databases which support the glob operator.

Again, thanks.
